cmd/src/debug_common.go

```diff
@@ -0,0 +1,470 @@
+package main
+
+import (
+	"archive/zip"
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"log"
+	"os"
+	"strings"
+	"sync"
+	"syscall"
+
+	"github.com/cockroachdb/errors"
+	"github.com/sourcegraph/src-cli/internal/exec"
+	"golang.org/x/sync/semaphore"
+)
+
+/*
+General Stuff
+TODO: file issue on the existence of OAuth signKey which needs to be redacted
+TODO: Create getSiteConfig function
+*/
+
+type archiveFile struct {
+	name string
+	data []byte
+	err  error
+}
+
+// setOpenFileLimits increases the limit of open files to the given number. This is needed
+// when doings lots of concurrent network requests which establish open sockets.
+func setOpenFileLimits(n uint64) error {
```

Comment:
Is this necessary now that we have a limited number of goroutines running at one time?

Reply:
I'll need to implement the semaphores on `src debug comp` and then try this out. Initially, this was added because the zip writer was opening up too many files at once, _I think_ that happens after the channel has been fully populated in closed but I'll need to have a closer look and test it.

Reply:
zip writer produces a serial stream of bytes; I'm guessing that the limit was there when you were _opening_ a lot of network requests (also files, weirdly!) but not writing individual files

Reply:
this was safely removed and tested against a kubernetes instance with 33 pods
```
warrengifford@Warrens-MacBook-Pro src-debug-test % src debug kube -n ns-sourcegraph -o final  
Archiving kubectl data for 33 pods
 SRC_ENDPOINT: https://cse-aws-test.sgdev.org
 Context: gke_beatrix-test-overlay_us-central1-c_beatrix-test
 Namespace: ns-sourcegraph
 Output filename: final.zip
Do you want to start writing to an archive? [y/N]: y
warrengifford@Warrens-MacBook-Pro src-debug-test % clear
warrengifford@Warrens-MacBook-Pro src-debug-test % unzip final.zip 
Archive:  final.zip
  inflating: final/kubectl/events.txt  
  inflating: final/kubectl/getPods.txt  
  inflating: final/kubectl/pods/grafana-0/grafana.log  
  inflating: final/kubectl/pods/codeintel-db-8fb75ddfd-xhncd/pgsql.log  
  inflating: final/kubectl/pods/codeinsights-db-78b8747cdd-bzj72/pgsql-exporter.log  
  inflating: final/kubectl/persistent-volumes.txt  
  inflating: final/kubectl/pods/cadvisor-44nvf/cadvisor.log  
  inflating: final/kubectl/pods/cadvisor-9dxct/cadvisor.log  
  inflating: final/kubectl/pods/redis-store-75ffcfb74b-b8pxf/redis-exporter.log  
  inflating: final/kubectl/pods/cadvisor-79pgg/cadvisor.log  
  inflating: final/kubectl/pods/cadvisor-gnnbq/cadvisor.log  
  inflating: final/kubectl/pods/cadvisor-cwmpm/cadvisor.log  
  inflating: final/kubectl/pods/github-proxy-7bfd897445-tgtxv/jaeger-agent.log  
  inflating: final/kubectl/pods/cadvisor-lmd6q/cadvisor.log  
  inflating: final/kubectl/pods/cadvisor-55b9d/cadvisor.log  
  inflating: final/kubectl/pods/cadvisor-nq75p/cadvisor.log  
  inflating: final/kubectl/pods/codeintel-db-8fb75ddfd-xhncd/pgsql-exporter.log  
  inflating: final/kubectl/pods/github-proxy-7bfd897445-tgtxv/github-proxy.log  
  inflating: final/kubectl/pods/cadvisor-zkspf/cadvisor.log  
  inflating: final/kubectl/persistent-volume-claims.txt  
  inflating: final/kubectl/pods/codeinsights-db-78b8747cdd-bzj72/codeinsights.log  
  inflating: final/kubectl/pods/gitserver-0/jaeger-agent.log  
  inflating: final/kubectl/pods/gitserver-0/gitserver.log  
  inflating: final/kubectl/pods/sourcegraph-frontend-6869dbb7d-tr7rk/jaeger-agent.log  
  inflating: final/kubectl/pods/sourcegraph-frontend-6869dbb7d-tr7rk/frontend.log  
  inflating: final/kubectl/pods/searcher-67fd7f64cc-l6dqs/jaeger-agent.log  
  inflating: final/kubectl/pods/pgsql-8565fd54ff-6n8tk/pgsql.log  
  inflating: final/kubectl/pods/sourcegraph-frontend-6869dbb7d-tf4v8/jaeger-agent.log  
  inflating: final/kubectl/pods/searcher-67fd7f64cc-l6dqs/searcher.log  
  inflating: final/kubectl/pods/repo-updater-978f46dff-27b52/repo-updater.log  
  inflating: final/kubectl/pods/pgsql-8565fd54ff-6n8tk/pgsql-exporter.log  
  inflating: final/kubectl/pods/repo-updater-978f46dff-27b52/jaeger-agent.log  
  inflating: final/kubectl/pods/cadvisor-7s8gw/cadvisor.log  
  inflating: final/kubectl/pods/precise-code-intel-worker-67946948c5-f9282/precise-code-intel-worker.log  
  inflating: final/kubectl/pods/precise-code-intel-worker-67946948c5-xq8pf/precise-code-intel-worker.log  
  inflating: final/kubectl/pods/indexed-search-0/zoekt-indexserver.log  
  inflating: final/kubectl/pods/jaeger-56fbb74f8-jmtjs/jaeger.log  
  inflating: final/kubectl/pods/indexed-search-1/zoekt-indexserver.log  
  inflating: final/kubectl/pods/minio-54b558cdd6-hlgbt/minio.log  
  inflating: final/kubectl/pods/symbols-95d9c59d9-v6brz/symbols.log  
  inflating: final/kubectl/pods/indexed-search-1/zoekt-webserver.log  
  inflating: final/kubectl/pods/indexed-search-0/zoekt-webserver.log  
  inflating: final/kubectl/pods/symbols-95d9c59d9-v6brz/jaeger-agent.log  
  inflating: final/kubectl/pods/searcher-67fd7f64cc-tpjj9/searcher.log  
  inflating: final/kubectl/pods/searcher-67fd7f64cc-tpjj9/jaeger-agent.log  
  inflating: final/kubectl/pods/worker-6dc84c8cf5-bkmdh/worker.log  
  inflating: final/kubectl/pods/gitserver-0/prev-gitserver.log  
  inflating: final/kubectl/pods/syntect-server-6b4478987-tpvsg/syntect-server.log  
  inflating: final/kubectl/pods/redis-cache-bb84b65cf-j6gfj/redis-exporter.log  
  inflating: final/kubectl/pods/prometheus-79f68b8d76-x8skc/prometheus.log  
  inflating: final/kubectl/pods/indexed-search-0/describe-indexed-search-0.txt  
  inflating: final/kubectl/pods/redis-store-75ffcfb74b-b8pxf/redis-store.log  
  inflating: final/kubectl/pods/cadvisor-gnnbq/describe-cadvisor-gnnbq.txt  
  inflating: final/kubectl/pods/jaeger-56fbb74f8-jmtjs/describe-jaeger-56fbb74f8-jmtjs.txt  
  inflating: final/kubectl/pods/redis-cache-bb84b65cf-j6gfj/redis-cache.log  
  inflating: final/kubectl/pods/indexed-search-1/describe-indexed-search-1.txt  
  inflating: final/kubectl/pods/cadvisor-cwmpm/describe-cadvisor-cwmpm.txt  
  inflating: final/kubectl/pods/cadvisor-lmd6q/describe-cadvisor-lmd6q.txt  
  inflating: final/kubectl/pods/grafana-0/manifest-grafana-0.yaml  
  inflating: final/kubectl/pods/cadvisor-nq75p/describe-cadvisor-nq75p.txt  
  inflating: final/kubectl/pods/repo-updater-978f46dff-27b52/prev-repo-updater.log  
  inflating: final/kubectl/pods/cadvisor-zkspf/describe-cadvisor-zkspf.txt  
  inflating: final/kubectl/pods/cadvisor-7s8gw/manifest-cadvisor-7s8gw.yaml  
  inflating: final/kubectl/pods/codeinsights-db-78b8747cdd-bzj72/describe-codeinsights-db-78b8747cdd-bzj72.txt  
  inflating: final/kubectl/pods/github-proxy-7bfd897445-tgtxv/describe-github-proxy-7bfd897445-tgtxv.txt  
  inflating: final/kubectl/pods/searcher-67fd7f64cc-tpjj9/describe-searcher-67fd7f64cc-tpjj9.txt  
  inflating: final/kubectl/pods/searcher-67fd7f64cc-l6dqs/describe-searcher-67fd7f64cc-l6dqs.txt  
  inflating: final/kubectl/pods/codeintel-db-8fb75ddfd-xhncd/describe-codeintel-db-8fb75ddfd-xhncd.txt  
  inflating: final/kubectl/pods/precise-code-intel-worker-67946948c5-xq8pf/describe-precise-code-intel-worker-67946948c5-xq8pf.txt  
  inflating: final/kubectl/pods/prometheus-79f68b8d76-x8skc/describe-prometheus-79f68b8d76-x8skc.txt  
  inflating: final/kubectl/pods/searcher-67fd7f64cc-l6dqs/prev-searcher.log  
  inflating: final/kubectl/pods/sourcegraph-frontend-6869dbb7d-tf4v8/describe-sourcegraph-frontend-6869dbb7d-tf4v8.txt  
  inflating: final/kubectl/pods/worker-6dc84c8cf5-bkmdh/prev-worker.log  
  inflating: final/kubectl/pods/minio-54b558cdd6-hlgbt/describe-minio-54b558cdd6-hlgbt.txt  
  inflating: final/kubectl/pods/redis-cache-bb84b65cf-j6gfj/describe-redis-cache-bb84b65cf-j6gfj.txt  
  inflating: final/kubectl/pods/precise-code-intel-worker-67946948c5-f9282/describe-precise-code-intel-worker-67946948c5-f9282.txt  
  inflating: final/kubectl/pods/cadvisor-44nvf/describe-cadvisor-44nvf.txt  
  inflating: final/kubectl/pods/cadvisor-55b9d/describe-cadvisor-55b9d.txt  
  inflating: final/kubectl/pods/cadvisor-79pgg/describe-cadvisor-79pgg.txt  
  inflating: final/kubectl/pods/pgsql-8565fd54ff-6n8tk/describe-pgsql-8565fd54ff-6n8tk.txt  
  inflating: final/kubectl/pods/cadvisor-7s8gw/describe-cadvisor-7s8gw.txt  
  inflating: final/kubectl/pods/gitserver-0/describe-gitserver-0.txt  
  inflating: final/kubectl/pods/cadvisor-zkspf/manifest-cadvisor-zkspf.yaml  
  inflating: final/kubectl/pods/codeintel-db-8fb75ddfd-xhncd/manifest-codeintel-db-8fb75ddfd-xhncd.yaml  
  inflating: final/kubectl/pods/cadvisor-9dxct/describe-cadvisor-9dxct.txt  
  inflating: final/kubectl/pods/repo-updater-978f46dff-27b52/describe-repo-updater-978f46dff-27b52.txt  
  inflating: final/kubectl/pods/cadvisor-cwmpm/manifest-cadvisor-cwmpm.yaml  
  inflating: final/kubectl/pods/cadvisor-44nvf/manifest-cadvisor-44nvf.yaml  
  inflating: final/kubectl/pods/codeinsights-db-78b8747cdd-bzj72/manifest-codeinsights-db-78b8747cdd-bzj72.yaml  
  inflating: final/kubectl/pods/cadvisor-gnnbq/manifest-cadvisor-gnnbq.yaml  
  inflating: final/kubectl/pods/sourcegraph-frontend-6869dbb7d-tr7rk/describe-sourcegraph-frontend-6869dbb7d-tr7rk.txt  
  inflating: final/kubectl/pods/cadvisor-55b9d/manifest-cadvisor-55b9d.yaml  
  inflating: final/kubectl/pods/cadvisor-79pgg/manifest-cadvisor-79pgg.yaml  
  inflating: final/kubectl/pods/worker-6dc84c8cf5-bkmdh/describe-worker-6dc84c8cf5-bkmdh.txt  
  inflating: final/kubectl/pods/cadvisor-nq75p/manifest-cadvisor-nq75p.yaml  
  inflating: final/kubectl/pods/syntect-server-6b4478987-tpvsg/describe-syntect-server-6b4478987-tpvsg.txt  
  inflating: final/kubectl/pods/indexed-search-0/manifest-indexed-search-0.yaml  
  inflating: final/kubectl/pods/redis-store-75ffcfb74b-b8pxf/describe-redis-store-75ffcfb74b-b8pxf.txt  
  inflating: final/kubectl/pods/precise-code-intel-worker-67946948c5-f9282/prev-precise-code-intel-worker.log  
  inflating: final/kubectl/pods/github-proxy-7bfd897445-tgtxv/manifest-github-proxy-7bfd897445-tgtxv.yaml  
  inflating: final/kubectl/pods/indexed-search-1/manifest-indexed-search-1.yaml  
  inflating: final/kubectl/pods/cadvisor-lmd6q/manifest-cadvisor-lmd6q.yaml  
  inflating: final/kubectl/pods/symbols-95d9c59d9-v6brz/describe-symbols-95d9c59d9-v6brz.txt  
  inflating: final/kubectl/pods/pgsql-8565fd54ff-6n8tk/prev-pgsql.log  
  inflating: final/kubectl/pods/pgsql-8565fd54ff-6n8tk/manifest-pgsql-8565fd54ff-6n8tk.yaml  
  inflating: final/kubectl/pods/searcher-67fd7f64cc-tpjj9/prev-searcher.log  
  inflating: final/kubectl/pods/repo-updater-978f46dff-27b52/manifest-repo-updater-978f46dff-27b52.yaml  
  inflating: final/kubectl/pods/sourcegraph-frontend-6869dbb7d-tr7rk/manifest-sourcegraph-frontend-6869dbb7d-tr7rk.yaml  
  inflating: final/kubectl/pods/gitserver-0/manifest-gitserver-0.yaml  
  inflating: final/kubectl/pods/searcher-67fd7f64cc-l6dqs/manifest-searcher-67fd7f64cc-l6dqs.yaml  
  inflating: final/kubectl/pods/jaeger-56fbb74f8-jmtjs/manifest-jaeger-56fbb74f8-jmtjs.yaml  
  inflating: final/kubectl/pods/searcher-67fd7f64cc-tpjj9/manifest-searcher-67fd7f64cc-tpjj9.yaml  
  inflating: final/kubectl/pods/minio-54b558cdd6-hlgbt/manifest-minio-54b558cdd6-hlgbt.yaml  
  inflating: final/kubectl/pods/redis-store-75ffcfb74b-b8pxf/manifest-redis-store-75ffcfb74b-b8pxf.yaml  
  inflating: final/kubectl/pods/redis-cache-bb84b65cf-j6gfj/manifest-redis-cache-bb84b65cf-j6gfj.yaml  
  inflating: final/kubectl/pods/sourcegraph-frontend-6869dbb7d-tf4v8/manifest-sourcegraph-frontend-6869dbb7d-tf4v8.yaml  
  inflating: final/kubectl/pods/worker-6dc84c8cf5-bkmdh/manifest-worker-6dc84c8cf5-bkmdh.yaml  
  inflating: final/kubectl/pods/precise-code-intel-worker-67946948c5-f9282/manifest-precise-code-intel-worker-67946948c5-f9282.yaml  
  inflating: final/kubectl/pods/precise-code-intel-worker-67946948c5-xq8pf/manifest-precise-code-intel-worker-67946948c5-xq8pf.yaml  
  inflating: final/kubectl/pods/symbols-95d9c59d9-v6brz/manifest-symbols-95d9c59d9-v6brz.yaml  
  inflating: final/kubectl/pods/syntect-server-6b4478987-tpvsg/manifest-syntect-server-6b4478987-tpvsg.yaml  
  inflating: final/kubectl/pods/cadvisor-9dxct/manifest-cadvisor-9dxct.yaml  
  inflating: final/kubectl/pods/prometheus-79f68b8d76-x8skc/manifest-prometheus-79f68b8d76-x8skc.yaml  
  inflating: final/kubectl/pods/precise-code-intel-worker-67946948c5-xq8pf/prev-precise-code-intel-worker.log  
  inflating: final/config/siteConfig.json  
  inflating: final/config/external_services.txt  
  inflating: final/kubectl/pods/grafana-0/describe-grafana-0.txt  
  inflating: final/kubectl/pods/jaeger-56fbb74f8-jmtjs/prev-jaeger.log  
  inflating: final/kubectl/pods/sourcegraph-frontend-6869dbb7d-tf4v8/frontend.log  

```